File: prettier-plugin-csharpier/src/Printer/Types/Interpolation.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { printSyntaxToken, SyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface InterpolationNode extends SyntaxTreeNode<"Interpolation"> {
    expression?: SyntaxTreeNode;
    alignmentClause?: InterpolationAlignmentClauseNode;
    formatClause?: InterpolationFormatClauseNode;
}

interface InterpolationAlignmentClauseNode extends SyntaxTreeNode<"InterpolationAlignmentClause"> {
    commaToken?: SyntaxToken;
    value?: SyntaxTreeNode;
}

interface InterpolationFormatClauseNode extends SyntaxTreeNode<"InterpolationFormatClause"> {
    colonToken?: SyntaxToken;
    formatStringToken?: SyntaxToken;
}

export const printInterpolation: PrintMethod<InterpolationNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = ["{", path.call(print, "expression")];
    if (node.alignmentClause) {
        parts.push(", ", path.call(print, "alignmentClause", "value"));
    }
    if (node.formatClause) {
        parts.push(":", printSyntaxToken(node.formatClause.formatStringToken));
    }
    parts.push("}");
    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/Interpolation.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class Interpolation
{
    public static Doc Print(InterpolationSyntax node, PrintingContext context)
    {
        var docs = new ValueListBuilder<Doc>([null, null, null, null, null, null, null]);
        docs.Append(Token.Print(node.OpenBraceToken, context));
        docs.Append(Node.Print(node.Expression, context));

        if (node.AlignmentClause != null)
        {
            docs.Append(
                Token.PrintWithSuffix(node.AlignmentClause.CommaToken, " ", context),
                Node.Print(node.AlignmentClause.Value, context)
            );
        }
        if (node.FormatClause != null)
        {
            docs.Append(
                Token.Print(node.FormatClause.ColonToken, context),
                Token.Print(node.FormatClause.FormatStringToken, context)
            );
        }

        docs.Append(Token.Print(node.CloseBraceToken, context));
        return Doc.Concat(ref docs);
    }
}
